Search for versioned cmdline tools instead of always using latest in AndroidSDK

## Steps to Reproduce

1. Run `flutter doctor --android-licenses` with Java 11 and (new) Android SDK command-line tools (new `sdkmanager`, etc.)

## Issue
Instead of using the command-line tools 3.0 installed on the machine, `flutter` sticks to the obsolete tools which are Java-8 only.

## Logs

<details>
<summary>Logs</summary>

```
$ sdkmanager --list_installed      
  ...                     
  cmdline-tools;3.0                                  | 3.0     | Android SDK Command-line Tools          | cmdline-tools/3.0/                                 
  ...
```

```
$ flutter doctor -v
[✓] Flutter (Channel stable, 1.22.6, on Linux, locale en_US.UTF-8)
    • Flutter version 1.22.6 at /home/saschpe/Applications/flutter
    • Framework revision 9b2d32b605 (4 weeks ago), 2021-01-22 14:36:39 -0800
    • Engine revision 2f0af37152
    • Dart version 2.10.5

[!] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /home/saschpe/.android/sdk
    • Platform android-30, build-tools 30.0.3
    • ANDROID_SDK_ROOT = /home/saschpe/.android/sdk
    • Java binary at: /usr/lib/jvm/java/bin/java
    • Java version OpenJDK Runtime Environment 18.9 (build 11.0.10+9)
    ✗ Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK licenses.
      See https://flutter.dev/docs/get-started/install/linux#android-setup for more details.

[!] Android Studio (not installed)
    • Android Studio not found; download from https://developer.android.com/studio/index.html
      (or visit https://flutter.dev/docs/get-started/install/linux#android-setup for detailed instructions).

[✓] Connected device (1 available)
    • sdk gphone x86 (mobile) • emulator-5554 • android-x86 • Android 11 (API 30) (emulator)

! Doctor found issues in 2 categories.
```
```
flutter doctor --android-licenses
Exception in thread "main" java.lang.NoClassDefFoundError: javax/xml/bind/annotation/XmlSchema
	at com.android.repository.api.SchemaModule$SchemaModuleVersion.<init>(SchemaModule.java:156)
	at com.android.repository.api.SchemaModule.<init>(SchemaModule.java:75)
	at com.android.sdklib.repository.AndroidSdkHandler.<clinit>(AndroidSdkHandler.java:81)
	at com.android.sdklib.tool.sdkmanager.SdkManagerCli.main(SdkManagerCli.java:73)
	at com.android.sdklib.tool.sdkmanager.SdkManagerCli.main(SdkManagerCli.java:48)
Caused by: java.lang.ClassNotFoundException: javax.xml.bind.annotation.XmlSchema
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:581)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
	... 5 more
```


</details>

--- a/packages/flutter_tools/lib/src/android/android_sdk.dart
+++ b/packages/flutter_tools/lib/src/android/android_sdk.dart
@@ -383,7 +383,24 @@ class AndroidSdk {
   ///
   /// The sdkmanager was previously in the tools directory but this component
   /// was marked as obsolete in 3.6.
-  String get sdkManagerPath => getCmdlineToolsPath(globals.platform.isWindows ? 'sdkmanager.bat': 'sdkmanager');
+  String get sdkManagerPath {
+    final String executable = globals.platform.isWindows
+      ? 'sdkmanager.bat'
+      : 'sdkmanager';
+    final File cmdlineTool = directory
+      .childDirectory('cmdline-tools')
+      .childDirectory('latest')
+      .childDirectory('bin')
+      .childFile(executable);
+    if (cmdlineTool.existsSync()) {
+      return cmdlineTool.path;
+    }
+    return directory
+      .childDirectory('tools')
+      .childDirectory('bin')
+      .childFile(executable)
+      .path;
+  }
 
   /// First try Java bundled with Android Studio, then sniff JAVA_HOME, then fallback to PATH.
   static String findJavaBinary({

--- a/packages/flutter_tools/lib/src/android/android_sdk.dart
+++ b/packages/flutter_tools/lib/src/android/android_sdk.dart
@@ -261,22 +261,55 @@ class AndroidSdk {
     return null;
   }
 
-  String getAvdManagerPath() {
-    final String binaryName = globals.platform.isWindows ? 'avdmanager.bat' : 'avdmanager';
-    final File cmdlineToolsBinary = directory
+  String getCmdlineToolsPath(String binaryName) {
+    // First look for the latest version of the command-line tools
+    final File cmdlineToolsLatestBinary = directory
       .childDirectory('cmdline-tools')
       .childDirectory('latest')
       .childDirectory('bin')
       .childFile(binaryName);
-    if (cmdlineToolsBinary.existsSync()) {
-      return cmdlineToolsBinary.path;
+    if (cmdlineToolsLatestBinary.existsSync()) {
+      return cmdlineToolsLatestBinary.path;
     }
+
+    // Next look for the highest version of the command-line tools
+    final Directory cmdlineToolsDir = directory.childDirectory('cmdline-tools');
+    if (cmdlineToolsDir.existsSync()) {
+      final List<Version> cmdlineTools = cmdlineToolsDir
+        .listSync()
+        .map((FileSystemEntity entity) {
+          try {
+            return Version.parse(entity.basename);
+          } on Exception {
+            return null;
+          }
+        })
+        .where((Version version) => version != null)
+        .toList();
+      cmdlineTools.sort();
+
+      for (final Version cmdlineToolsVersion in cmdlineTools.reversed) {
+        final File cmdlineToolsBinary = directory
+          .childDirectory('cmdline-tools')
+          .childDirectory(cmdlineToolsVersion.toString())
+          .childDirectory('bin')
+          .childFile(binaryName);
+        if (cmdlineToolsBinary.existsSync()) {
+          return cmdlineToolsBinary.path;
+        }
+      }
+    }
+
+    // Finally fallback to the old SDK tools
     final File toolsBinary = directory.childDirectory('tools').childDirectory('bin').childFile(binaryName);
     if (toolsBinary.existsSync()) {
       return toolsBinary.path;
     }
+
     return null;
   }
+
+  String getAvdManagerPath() => getCmdlineToolsPath(globals.platform.isWindows ? 'avdmanager.bat' : 'avdmanager');
 
   /// Sets up various paths used internally.
   ///
@@ -350,24 +383,7 @@ class AndroidSdk {
   ///
   /// The sdkmanager was previously in the tools directory but this component
   /// was marked as obsolete in 3.6.
-  String get sdkManagerPath {
-    final String executable = globals.platform.isWindows
-      ? 'sdkmanager.bat'
-      : 'sdkmanager';
-    final File cmdlineTool = directory
-      .childDirectory('cmdline-tools')
-      .childDirectory('latest')
-      .childDirectory('bin')
-      .childFile(executable);
-    if (cmdlineTool.existsSync()) {
-      return cmdlineTool.path;
-    }
-    return directory
-      .childDirectory('tools')
-      .childDirectory('bin')
-      .childFile(executable)
-      .path;
-  }
+  String get sdkManagerPath => getCmdlineToolsPath(globals.platform.isWindows ? 'sdkmanager.bat': 'sdkmanager');
 
   /// First try Java bundled with Android Studio, then sniff JAVA_HOME, then fallback to PATH.
   static String findJavaBinary({

--- a/packages/flutter_tools/lib/src/android/android_sdk.dart
+++ b/packages/flutter_tools/lib/src/android/android_sdk.dart
@@ -387,14 +387,11 @@ class AndroidSdk {
     final String executable = globals.platform.isWindows
       ? 'sdkmanager.bat'
       : 'sdkmanager';
-    final File cmdlineTool = directory
-      .childDirectory('cmdline-tools')
-      .childDirectory('latest')
-      .childDirectory('bin')
-      .childFile(executable);
-    if (cmdlineTool.existsSync()) {
-      return cmdlineTool.path;
+    final String path = getCmdlineToolsPath(executable);
+    if (path != null) {
+      return path;
     }
+    // If no binary was found, return the default location
     return directory
       .childDirectory('tools')
       .childDirectory('bin')

--- a/packages/flutter_tools/lib/src/android/android_sdk.dart
+++ b/packages/flutter_tools/lib/src/android/android_sdk.dart
@@ -277,9 +277,10 @@ class AndroidSdk {
     if (cmdlineToolsDir.existsSync()) {
       final List<Version> cmdlineTools = cmdlineToolsDir
         .listSync()
-        .map((FileSystemEntity entity) {
+        .whereType<Directory>()
+        .map((Directory subDirectory) {
           try {
-            return Version.parse(entity.basename);
+            return Version.parse(subDirectory.basename);
           } on Exception {
             return null;
           }

--- a/packages/flutter_tools/test/general.shard/android/android_sdk_test.dart
+++ b/packages/flutter_tools/test/general.shard/android/android_sdk_test.dart
@@ -80,6 +80,26 @@ void main() {
       Config: () => config,
     });
 
+    testUsingContext('returns sdkmanager path under cmdline tools (highest version) on Linux/macOS', () {
+      sdkDir = MockAndroidSdk.createSdkDirectory();
+      config.setValue('android-sdk', sdkDir.path);
+
+      final AndroidSdk sdk = AndroidSdk.locateAndroidSdk();
+      final List<String> versions = <String>['3.0', '2.1', '1.0'];
+      for (final String version in versions) {
+        fileSystem.file(
+          fileSystem.path.join(sdk.directory.path, 'cmdline-tools', version, 'bin', 'sdkmanager')
+        ).createSync(recursive: true);
+      }
+
+      expect(sdk.sdkManagerPath, fileSystem.path.join(sdk.directory.path, 'cmdline-tools', '3.0', 'bin', 'sdkmanager'));
+    }, overrides: <Type, Generator>{
+      FileSystem: () => fileSystem,
+      ProcessManager: () => FakeProcessManager.any(),
+      Platform: () => FakePlatform(operatingSystem: 'linux'),
+      Config: () => config,
+    });
+
     testUsingContext('Caches adb location after first access', () {
       sdkDir = MockAndroidSdk.createSdkDirectory();
       config.setValue('android-sdk', sdkDir.path);

